### ISSUE TITLE
Trailing backslash reruns makes a redirect that looses https.

### ISSUE THREAD

**Issue by atompie** [tiangolo/fastapi#4990]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
application = FastAPI()

application.add_middleware(
    CORSMiddleware,
    allow_origins=["*"],
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)
```

### Description

I am running standard FastAPI with CORS middleware. I Noticed that a trailing backslash at the URL makes a 307 Temporary Redirect. And it is fine when you have an http URL. But it is not working as expected when you have a HTTPS URL. It replaces HTTPS with HTTP. 

![https_http](https://user-images.githubusercontent.com/16271564/172128733-90e31032-ab82-417f-be1f-14754deadaac.png)

This is quite hard to catch and causes a lot of confusion why the endpoint is not responding.

![time line](https://user-images.githubusercontent.com/16271564/172129162-70fd5686-d28b-4f09-93ca-04fb82d548e7.png)

### Operating System

Linux

### FastAPI Version

0.78.0

### Python Version

3.8.10

### Additional Context

This may be important. The SSL certificate on this page is a self-signed cert.

**Comment by anmolmittal1** [comment]
@JarroVGIT Yes, I am facing the same issue and i am running fastapi behind traefik (reverse proxy)

**Comment by JarroVGIT** [comment]
I am thinking that the SSL is terminated at your proxy, and so the actual request that hits FastAPI is already without SSL. The redirect happens in Starlette code base (routing,py -> Router -> __call__()), and it doesn't remove https. 

Uvicorn has a piece of info in their docs, it is about Nginx but it would be similar to Traefik. https://www.uvicorn.org/deployment/#running-behind-nginx

Hope that helps solving this issue.

**Comment by irv** [comment]
This does appear to be a legit issue, which I think can be demonstrated as follows:

```python
@app.head("/path/")
  def example_path(): -> PlainTextResponse
   return responses.PlainTextResponse("Example")
```

1) When running locally, and binding uvicorn to localhost, it works as expected:

```
uvicorn server:app --host 127.0.0.1 --port 8001

curl -I -H "X-Forwarded-Proto: https" -H "Host: blah.com" -H "X-Forwarded-For: 169.255.255.1" 127.0.0.1:8001/path
HTTP/1.1 307 Temporary Redirect
date: Thu, 19 Jan 2023 10:37:58 GMT
server: uvicorn
location: https://blah.com/path/
```

2) When running locally, and binding uvicorn to my network address, it fails in the way described in this issue:
```
uvicorn server:app --host 192.168.1.79 --port 8001

curl -I -H "X-Forwarded-Proto: https" -H "Host: blah.com" -H "X-Forwarded-For: 169.255.255.1" 192.168.1.79:8001/path
HTTP/1.1 307 Temporary Redirect
date: Thu, 19 Jan 2023 10:39:23 GMT
server: uvicorn
location: http://blah.com/path/
```

I haven't tested _all_ the various `X-Forwarded-*` headers, but `X-Forwarded-Proto` is definitely ignored in this trailing slash case, and seemingly _only_ in this trailing slash case

**Comment by atompie** [comment]
> Are you running a reverse proxy? Or are you testing this on localhost?

On localhost
